src/Sentry/Internal/Http/CachingTransport.cs

```diff
@@ -0,0 +1,266 @@
+                    Directory.CreateDirectory(_cacheDirectoryPath);
+
+                    File.Move(
+                        envelopeFilePath,
+                        Path.Combine(_cacheDirectoryPath, Path.GetFileName(envelopeFilePath))
+                    );
+
+                    // Avoid processing other files because the order may end up corrupted
+                    break;
+                }
+                else
+                {
+                    // Delete the envelope file and move on to the next one
+                    File.Delete(envelopeFilePath);
+                }
+            }
+        }
+
+        private async ValueTask StoreToCacheAsync(
+            Envelope envelope,
+            CancellationToken cancellationToken = default)
+        {
+            using var lockClaim = await _cacheDirectoryLock.ClaimAsync(cancellationToken).ConfigureAwait(false);
+
+            // If over capacity - remove oldest envelope file
+            while (
+                GetCacheFilePaths().Count() >= _options.MaxQueueItems &&
+                GetCacheFilePaths().FirstOrDefault() is { } oldestEnvelopeFilePath)
+            {
+                File.Delete(oldestEnvelopeFilePath);
+            }
+
+            // Envelope file name can be either:
+            // 1604679692_b2495755f67e4bb8a75504e5ce91d6c1_17754019.envelope
+            // 1604679692__17754019.envelope
+            // (depending on whether event ID is present or not)
+            var envelopeFilePath = Path.Combine(
+                _cacheDirectoryPath,
+                $"{DateTimeOffset.UtcNow.ToUnixTimeSeconds()}_" +
+                $"{envelope.TryGetEventId()}_" +
+                $"{envelope.GetHashCode()}" +
+                $".{EnvelopeFileExt}"
+            );
+
+            Directory.CreateDirectory(_cacheDirectoryPath);
+
+            using (var stream = File.Create(envelopeFilePath))
+            {
+                await envelope.SerializeAsync(stream, cancellationToken).ConfigureAwait(false);
+            }
+
+            // Tell the worker that there is work available
+            // (file stream MUST BE DISPOSED prior to this)
+            _workerSignal.Release();
+        }
+
+        // This method asynchronously blocks until the envelope is written to cache, but not until it's sent
+        public async ValueTask SendEnvelopeAsync(
+            Envelope envelope,
+            CancellationToken cancellationToken = default)
+        {
+            // Store the envelope in a file without actually sending it anywhere.
+            // The envelope will get picked up by the background thread eventually.
+            await StoreToCacheAsync(envelope, cancellationToken).ConfigureAwait(false);
+        }
+
+        public async ValueTask StopWorkerAsync()
+        {
+            // Stop worker and wait until it finishes
+            _workerCts.Cancel();
+            await _worker.ConfigureAwait(false);
```

Comment:
Since we call this from the `Dispose` method, awaiting on a Task that fails throws. Should we try/catch here?